### ISSUE TITLE
Support for Kiwi Browser on Android

### ISSUE THREAD

**Issue by Sekers** [xbrowsersync/app#126]
Kiwi browser is becoming an amazing Chrome alternative that still uses Chromium but has built-in ad blocker, can bypass AMP links, etc. Recently, they added Chrome extension support which is pretty amazing. For example, the Bitwarden Chrome extension works perfectly.

I have been having trouble getting the xBrowserSynx Chrome extension to work in Kiwi. Would it be possible to get support of the Kiwi browser added to the roadmap?

**Comment by nero120** [comment]
Thanks @Sekers, I don't know much about Kiwi but once I start work on v1.6.0 (next release after the current one I'm working on, which will include support for other Chromium browsers) I'll look into how much effort it will be to support Kiwi as well.

**Comment by Sekers** [comment]
Thank you. Some have gotten it the extension to work as recently as a few weeks ago, but something broke since then. I am hoping this means that it's not a huge amount of work to have the two projects work together.

Much appreciated!

**Comment by nero120** [comment]
Hopefully not! 🤞

**Comment by nero120** [comment]
Totally didn't realise this was a mobile only browser! When you said chromium I just assumed it was desktop... 😖 

xBrowserSync currently does not support integration with any mobile browser (to access your synced bookmarks on mobile you'd need to use the xBrowserSync android app). It is not currently on the [roadmap](https://github.com/xbrowsersync/app/wiki/Roadmap) and is unlikely to change for the foreseeable future, really sorry.

**Comment by rodrigoswz** [comment]
> Totally didn't realise this was a mobile only browser! When you said chromium I just assumed it was desktop... 😖 
> 
> xBrowserSync currently does not support integration with any mobile browser (to access your synced bookmarks on mobile you'd need to use the xBrowserSync android app). It is not currently on the [roadmap](https://github.com/xbrowsersync/app/wiki/Roadmap) and is unlikely to change for the foreseeable future, really sorry.

Sorry to discuss this issue closed, I would like to inform that Kiwi so far is the only browser that supports Chrome extensions on mobile. [Btw it's recently fully open source](https://www.xda-developers.com/kiwi-browser-open-source/).

[Until version 1.4.0 of xBrowserSync this worked perfectly](https://i.imgur.com/ioCcIn1.jpg), in the [new updates (1.5.x) this stopped working](https://i.imgur.com/pwsYdY5.jpg).

It would be possible to analyze what happened to break the extension's functionality in Kiwi?

**Comment by nero120** [comment]
@rodrigoswz I'm currently working on an [API](https://github.com/xbrowsersync/api) release, once that's done and I start work on the next app release I will try to see why xBrowserSync stopped working in Kiwi and fix the issue for the next release.

**Comment by remipp** [comment]
@nero120 in case you're still interested in fixing this: I tested a bit and it fails and throws an exception in chromium-bookmark.service.js in getNativeContainerIds() because on mobile there is no "other bookmarks" or "bookmarks toolbar". There is only "Mobile bookmarks" (which is just part of tree[0] like the others and not under "Other bookmarks" like the code currently thinks (line 430)). It worked in v1.4.0 simply because it did less error handling and didn't throw an exception there.
I got it to work by not throwing if "Mobile bookmarks" exists and doing the same in the clearing/population code. However I'm not really a javascript developer and I don't feel like this hacky solution is ready for a PR yet.

**Comment by nero120** [comment]
@remipp thanks for the useful info! If that's the case kiwi will need it's own platform implementation - if you currently try to install xBrowserSync on kiwi it will be using the chrome/chromium implementation which expects both Other bookmarks and Bookmarks bar to exist.

This is fine, however as kiwi is not a major browser it's not the highest of priorities so it goes to the back of the queue behind Opera and Vivaldi. Depending on how long it takes to implement those I will then look at implementing other platforms based on popularity.

**Comment by Gromchy** [comment]
sorry for asking, but out of curiosity, is there any update on this?
XbrowserSync is installed, I login fine, but it keeps saying "syncing" but never actually finishes.

On my Google Chrome desktop its been confirmed as working fine. But syncing to Kiwi mobile is still an issue for me.

**Comment by qol3m** [comment]
Same problem here. "syncing", but nothing happens.

**Comment by michkot** [comment]
I am myself Opera & Kiwi user, and it seemed like I could figure out how to properly extend your @nero120 extension, so I gave it a try. I started with a basic hack that @remipp pointed to... and then I went on and refactored a lot of code... 
https://github.com/xbrowsersync/app/compare/master...michkot:master

TL;DR it works well with Opera@Desktop and Kiwi@Android but I can not provide any guarantees guys - also, the store version is currently 1.5.2, I modified latest code at master with v1.6.0

If you feel like the changes I made are going in a good direction, let me know, I will open a pull request and we can discuss it more technically (I would polish it and probably "port" the refactoring over to the firefox platform too).

**Comment by nero120** [comment]
Thanks for this @michkot! I will try and encorporate your changes into v1.6.0. It'll get released eventually, just wish I had more free time! 😢

**Comment by Oxika95** [comment]
Came here to mention I am also interested in getting this up and running. Getting Kiwi support will bring me one step closer to cutting google out of the equation.

**Comment by mvbureev** [comment]
The problem is still relevant, but I have the opportunity to go into the extension and even specify id:password. But after entering and long loading, an error appears.
<img width="304" alt="image" src="https://user-images.githubusercontent.com/44510263/188368357-d84684ba-d17e-414b-b8d9-443f8fa325e4.png">

**Comment by simwai** [comment]
Please guys support kiwi browser. This browser is the best on Android and really deserves it to be supported. Otherwise, this project here is useless for a lot of people (like me).

**Comment by SPDurkee** [comment]
Why is this closed? It is definitely still and issue. Since kiwi is the only android browser that has full extension support, it may be worth the effort. Especially if it is just some additional error handling as mentioned in this issue.

**Comment by rokejulianlockhart** [comment]
> Since kiwi is the only android browser that has full extension support

https://www.androidpolice.com/firefox-bringing-back-full-extension-support-android/

I've been using full extension support on Firefox for 3 years.

**Comment by SPDurkee** [comment]
> @SPDurkee,
> 
> > Since kiwi is the only android browser that has full extension support
> 
> https://www.androidpolice.com/firefox-bringing-back-full-extension-support-android/
> 
> I've been using full extension support on Firefox for 3 years.

I assume you put it in a collection and are using it with Firefox nightly. Can you confirm that xbrowsersync is indeed working correctly and that configuration? Thanks!

**Comment by rokejulianlockhart** [comment]
> https://github.com/xbrowsersync/app/issues/126#issuecomment-1676163139

@SPDurkee, it does not.:

![image](https://github.com/xbrowsersync/app/assets/42837531/5f359652-1e93-48b1-ba15-e63467fc4495)
